deep_qa/training/train_utils.py

```diff
@@ -0,0 +1,113 @@
+from typing import List, Tuple
+import tensorflow
+
+
+def pin_variable_device_scope(device, variable_device="/cpu:0"):
+    """
+    Tensorflow device scopes can take functions which give a device
+    for a given op in the graph. Here, we use the device that is
+    passed to the scope *unless* the operation which is being created
+    in the graph is a Variable creation op; in this case, we place it
+    on the cpu.
+    """
+    def _assign(graph_op):
+        node_def = graph_op if isinstance(graph_op, tensorflow.NodeDef) else graph_op.node_def
+        if node_def.op in ['Variable', 'VariableV2']:
+            return variable_device
+        else:
+            return device
+    return _assign
+
+
+def average_gradients(tower_gradients: List[List[Tuple]]):
+    """
+    Given a list of (gradient, variable) pairs from the result of
+    a gradient calculation from multiple GPUs, calculate their
+    average.
+    """
+    # Make a map from variables -> [gradients that are not none].
+    gradient_map = {}
+    for tower in tower_gradients:
+        for grad, variable in tower:
+            if grad is not None:
+                if variable not in gradient_map:
+                    gradient_map[variable] = []
+                gradient_map[variable].append(grad)
+    average_gradient_list = []
+
+    for variable, gradients in gradient_map.items():
+        # variable is a tensor.
+        # gradients is a list of gradients for this tensor to average.
+
+        # Pick any one of the gradients to see if it is an IndexedSlice.
+
+        first_actual_grad = gradients[0]
+        if isinstance(first_actual_grad, tensorflow.IndexedSlices):
+            # If the gradient is type IndexedSlices then this is a sparse
+            # gradient with attributes indices and values. To average, we
+            # need to concat them individually and then create a new
+            # IndexedSlices object.
+            indices = []
+            values = []
+            for grad in gradients:
+                indices.append(grad.indices)
+                values.append(grad.values)
+            all_indices = tensorflow.concat(indices, 0)
+            avg_values = tensorflow.concat(values, 0) / len(gradients)
+            # Deduplicate across indices.
+            values, unique_indicies = _deduplicate_indexed_slices(avg_values, all_indices)
```

Comment:
Or, another option, just split this whole block into a `_average_sparse_gradients` method, and split the `else` block into a `_average_dense_gradients` method, which would give a simpler method, anyway.  Then the `_deduplicate_indexed_slices` method could just be in-lined.  If you go that route, you might think a bit more about names - there's a lot of overlap between `average_gradients` and the names I suggested.